File: src/v3/runtime/context.ts
import type { DiceEventSink } from './journal.js';
import { ExecutionJournal } from './journal.js';
import { ExecutionBudget } from './budget.js';
import { createDiceLimits, type DiceLimitOverrides, type DiceLimits } from './limits.js';
import { MersenneTwister19937, type RandomSource } from './mt19937.js';
import {
  createReplayDescriptor,
  createReplaySeed,
  createSeedMaterial,
  validateReplayDescriptor,
  type CryptoSource,
  type RandomAlgorithm,
  type ReplayDescriptor,
  type SeedInput,
  type SeedMaterial,
} from './replay.js';
import { Xoshiro128StarStar } from './xoshiro128ss.js';
import { DiceRollError } from '../errors.js';

export interface ExecutionContextOptions {
  readonly limits?: DiceLimits | DiceLimitOverrides;
  /** Already resolved immutable limits used by the engine hot path. */
  readonly resolvedLimits?: DiceLimits;
  readonly seed?: SeedInput;
  readonly replay?: ReplayDescriptor;
  readonly randomAlgorithm?: RandomAlgorithm;
  readonly planFingerprint?: string;
  readonly collectEvents?: boolean;
  /** Intended for platform adapters and deterministic tests. */
  readonly cryptoSource?: CryptoSource | null;
}

/** All mutable state for exactly one compilation/execution. */
export class ExecutionContext {
  readonly limits: DiceLimits;

  readonly budget: ExecutionBudget;

  readonly random: RandomSource;

  readonly journal: DiceEventSink;

  readonly replay: ReplayDescriptor;

  constructor(options: ExecutionContextOptions = {}) {
    this.limits = options.resolvedLimits ?? createDiceLimits(options.limits);
    this.budget = new ExecutionBudget(this.limits);
    if (options.seed !== undefined && options.replay !== undefined) {
      throw new DiceRollError('seed and replay are mutually exclusive', {
        code: 'INVALID_REPLAY',
      });
    }
    if (options.replay !== undefined && options.randomAlgorithm !== undefined) {
      throw new DiceRollError('A replay descriptor determines its random algorithm', {
        code: 'INVALID_REPLAY',
      });
    }
    if (options.collectEvents !== undefined && typeof options.collectEvents !== 'boolean') {
      throw new DiceRollError('collectEvents must be a boolean', {
        code: 'ROLL_EXECUTION_LIMIT',
      });
    }

    let algorithm: RandomAlgorithm;
    let seed: SeedMaterial;
    if (options.replay === undefined) {
      algorithm = options.randomAlgorithm ?? 'mt19937';
      if (algorithm !== 'mt19937' && algorithm !== 'xoshiro128ss') {
        throw new DiceRollError('The requested random algorithm is not supported', {
          code: 'INVALID_REPLAY',
        });
      }
      seed = createSeedMaterial(
        options.seed,
        options.cryptoSource,
        this.limits.maxSeedLength,
      );
      this.replay = createReplayDescriptor(seed, {
        algorithm,
        ...(options.planFingerprint === undefined
          ? {}
          : { planFingerprint: options.planFingerprint }),
      });
    } else {
      const replay = validateReplayDescriptor(options.replay, options.planFingerprint);
      algorithm = replay.algorithm;
      seed = createReplaySeed(replay, options.planFingerprint);
      this.replay = replay;
    }
    this.random = algorithm === 'mt19937'
      ? new MersenneTwister19937(seed.words, this.budget)
      : new Xoshiro128StarStar(seed.words, this.budget);
    this.journal = new ExecutionJournal(this.budget, options.collectEvents ?? true);
  }
}

export function createExecutionContext(options: ExecutionContextOptions = {}): ExecutionContext {
  return new ExecutionContext(options);
}

File: go/context.go
package dicecore

import "io"

// ExecutionContextOptions controls one execution's independent mutable state.
type ExecutionContextOptions struct {
	Limits               *DiceLimits
	LimitOverrides       DiceLimitOverrides
	ResolvedLimits       *DiceLimits
	Seed                 any
	Replay               any
	RandomAlgorithm      RandomAlgorithm
	PlanFingerprint      string
	CollectEvents        *bool
	CryptoSource         io.Reader
	CryptoSourceProvided bool
}

type ExecutionContext struct {
	Limits  DiceLimits
	Budget  *ExecutionBudget
	Random  RandomSource
	Journal *ExecutionJournal
	Replay  ReplayDescriptor
}

func CreateExecutionContext(options ...ExecutionContextOptions) (*ExecutionContext, error) {
	var option ExecutionContextOptions
	if len(options) > 0 {
		option = options[0]
	}
	limits := DefaultDiceLimits()
	if option.ResolvedLimits != nil {
		limits = *option.ResolvedLimits
	} else if option.Limits != nil {
		limits = *option.Limits
		if err := limits.Validate(); err != nil {
			return nil, err
		}
	} else {
		var err error
		limits, err = CreateDiceLimits(option.LimitOverrides)
		if err != nil {
			return nil, err
		}
	}
	if option.Seed != nil && option.Replay != nil {
		return nil, newDiceError("INVALID_REPLAY", "seed and replay are mutually exclusive", "", nil)
	}
	if option.Replay != nil && option.RandomAlgorithm != "" {
		return nil, newDiceError("INVALID_REPLAY", "A replay descriptor determines its random algorithm", "", nil)
	}
	context := &ExecutionContext{Limits: limits, Budget: NewExecutionBudget(limits)}
	var seed SeedMaterial
	var algorithm RandomAlgorithm
	var err error
	if option.Replay == nil {
		algorithm = option.RandomAlgorithm
		if algorithm == "" {
			algorithm = MT19937
		}
		if !supportedRandomAlgorithm(algorithm) {
			return nil, newDiceError("INVALID_REPLAY", "The requested random algorithm is not supported", "", nil)
		}
		if option.Seed != nil {
			seed, err = CreateProvidedSeed(option.Seed, limits.MaxSeedLength)
		} else if option.CryptoSourceProvided || option.CryptoSource != nil {
			seed, err = CreateAutomaticSeed(option.CryptoSource)
		} else {
			seed, err = CreateAutomaticSeed()
		}
		if err != nil {
			return nil, err
		}
		context.Replay, err = CreateReplayDescriptor(seed, ReplayDescriptorOptions{Algorithm: algorithm, PlanFingerprint: option.PlanFingerprint})
	} else {
		var restored ReplayState
		if option.PlanFingerprint != "" {
			restored, err = CreateReplayState(option.Replay, option.PlanFingerprint)
		} else {
			restored, err = CreateReplayState(option.Replay)
		}
		seed, context.Replay = restored.Seed, restored.Replay
		algorithm = restored.Replay.Algorithm
	}
	if err != nil {
		return nil, err
	}
	return initializeExecutionContextRandom(context, algorithm, seed.Words[:], option.CollectEvents)
}

// Initializing the generator is separate from resolving seed/replay material;
// generator validation errors leave no partially initialized context available.
func initializeExecutionContextRandom(context *ExecutionContext, algorithm RandomAlgorithm, words []uint32, collectEvents *bool) (*ExecutionContext, error) {
	var err error
	if algorithm == MT19937 {
		context.Random, err = NewMersenneTwister19937FromWords(words, context.Budget)
	} else {
		context.Random, err = NewXoshiro128StarStar(words, context.Budget)
	}
	if err != nil {
		return nil, err
	}
	materialize := true
	if collectEvents != nil {
		materialize = *collectEvents
	}
	context.Journal = NewExecutionJournal(context.Budget, materialize)
	return context, nil
}
